File: googleplay.go
package googleplay

import (
   "bytes"
   "crypto/rsa"
   "crypto/sha1"
   "encoding/base64"
   "encoding/json"
   "fmt"
   "github.com/89z/parse/crypto"
   "github.com/89z/parse/net"
   "github.com/segmentio/encoding/proto"
   "io"
   "math/big"
   "net/http"
   "net/url"
   "strconv"
   "strings"
)

const AndroidJA3 =
   "769,49195-49196-52393-49199-49200-52392-158-159-49161-49162-49171-49172-" +
   "51-57-156-157-47-53,65281-0-23-35-13-16-11-10,23,0"

const androidKey =
   "AAAAgMom/1a/v0lblO2Ubrt60J2gcuXSljGFQXgcyZWveWLEwo6prwgi3iJIZdodyhKZQrNWp" +
   "5nKJ3srRXcUW+F1BD3baEVGcmEgqaLZUNBjm057pKRI16kB0YppeGx5qIQ5QjKzsR8ETQbKLN" +
   "WgRY0QRNVz34kMJR3P/LgHax/6rmf5AAAAAwEAAQ=="

const origin = "https://android.clients.google.com"

func Signature(email, password string) (string, error) {
   data, err := base64.StdEncoding.DecodeString(androidKey)
   if err != nil {
      return "", err
   }
   var key rsa.PublicKey
   buf := crypto.NewBuffer(data)
   // modulus_length | modulus | exponent_length | exponent
   _, mod, ok := buf.ReadUint32LengthPrefixed()
   if ok {
      key.N = new(big.Int).SetBytes(mod)
   }
   _, exp, ok := buf.ReadUint32LengthPrefixed()
   if ok {
      exp := new(big.Int).SetBytes(exp).Int64()
      key.E = int(exp)
   }
   var (
      msg bytes.Buffer
      zero devZero
   )
   msg.WriteString(email)
   msg.WriteByte(0)
   msg.WriteString(password)
   login, err := rsa.EncryptOAEP(
      sha1.New(), zero, &key, msg.Bytes(), nil,
   )
   if err != nil {
      return "", err
   }
   hash := sha1.Sum(data)
   msg.Reset()
   msg.WriteByte(0)
   msg.Write(hash[:4])
   msg.Write(login)
   return base64.URLEncoding.EncodeToString(msg.Bytes()), nil
}

type Checkin struct {
   Checkin struct{} `json:"checkin"`
   Version int `json:"version"`
}

func NewCheckin() Checkin {
   return Checkin{Version: 3}
}

func (c Checkin) Post() (*CheckinResponse, error) {
   buf := new(bytes.Buffer)
   err := json.NewEncoder(buf).Encode(c)
   if err != nil {
      return nil, err
   }
   res, err := http.Post(origin + "/checkin", "application/json", buf)
   if err != nil {
      return nil, err
   }
   defer res.Body.Close()
   if res.StatusCode != http.StatusOK {
      return nil, fmt.Errorf("status %q", res.Status)
   }
   check := new(CheckinResponse)
   if err := json.NewDecoder(res.Body).Decode(check); err != nil {
      return nil, err
   }
   return check, nil
}

type CheckinResponse struct {
   Android_ID int64
}

func (c CheckinResponse) String() string {
   return strconv.FormatInt(c.Android_ID, 16)
}

type Device struct {
   Configuration struct {
      GlEsVersion            int32   `protobuf:"varint,8"`
      HasFiveWayNavigation   bool    `protobuf:"varint,6"`
      HasHardKeyboard        bool    `protobuf:"varint,5"`
      Keyboard               int32   `protobuf:"varint,2"`
      NativePlatform         []string `protobuf:"bytes,11"`
      Navigation             int32   `protobuf:"varint,3"`
      ScreenDensity          int32   `protobuf:"varint,7"`
      ScreenLayout           int32   `protobuf:"varint,4"`
      SystemAvailableFeature []string `protobuf:"bytes,10"`
      TouchScreen            int32   `protobuf:"varint,1"`
   } `protobuf:"bytes,1"`
}

func NewDevice() Device {
   var d Device
   // OpenGL ES version
   // developer.android.com/guide/topics/manifest/uses-feature-element
   d.Configuration.GlEsVersion = 0x0002_0000
   d.Configuration.HasFiveWayNavigation = true
   d.Configuration.HasHardKeyboard = true
   d.Configuration.Keyboard = 1
   d.Configuration.NativePlatform = []string{"armeabi-v7a"}
   d.Configuration.Navigation = 1
   d.Configuration.ScreenDensity = 1
   d.Configuration.ScreenLayout = 1
   d.Configuration.SystemAvailableFeature = []string{
      "android.hardware.touchscreen",
      "android.hardware.wifi",
   }
   d.Configuration.TouchScreen = 1
   return d
}

func (d Device) Upload(check *CheckinResponse, oauth OAuth) error {
   buf, err := proto.Marshal(d)
   if err != nil {
      return err
   }
   req, err := http.NewRequest(
      "POST", origin + "/fdfe/uploadDeviceConfig", bytes.NewReader(buf),
   )
   if err != nil {
      return err
   }
   req.Header = http.Header{
      "Authorization": {"Bearer " + oauth.Get("Auth")},
      "User-Agent": {"Android-Finsky (versionCode=81031200,sdk=27)"},
      "X-DFE-Device-Id": {check.String()},
   }
   res, err := new(http.Transport).RoundTrip(req)
   if err != nil {
      return err
   }
   defer res.Body.Close()
   if res.StatusCode != http.StatusOK {
      return fmt.Errorf("status %q", res.Status)
   }
   return nil
}

type OAuth struct {
   url.Values
}

// device is Google Service Framework.
func (a OAuth) Details(device, app string) ([]byte, error) {
   req, err := http.NewRequest("GET", origin + "/fdfe/details", nil)
   if err != nil {
      return nil, err
   }
   val := req.URL.Query()
   val.Set("doc", app)
   req.URL.RawQuery = val.Encode()
   req.Header.Set("Authorization", "Bearer " + a.Get("Auth"))
   req.Header.Set("X-DFE-Device-ID", device)
   res, err := new(http.Transport).RoundTrip(req)
   if err != nil {
      return nil, err
   }
   defer res.Body.Close()
   if res.StatusCode != http.StatusOK {
      return nil, fmt.Errorf("status %q", res.Status)
   }
   return io.ReadAll(res.Body)
}

type Token struct {
   url.Values
}

// Request refresh token.
func NewToken(email, password string) (*Token, error) {
   hello, err := crypto.ParseJA3(AndroidJA3)
   if err != nil {
      return nil, err
   }
   sig, err := Signature(email, password)
   if err != nil {
      return nil, err
   }
   val := make(url.Values)
   val.Set("Email", email)
   val.Set("EncryptedPasswd", sig)
   val.Set("sdk_version", "16")
   req, err := http.NewRequest(
      "POST", origin + "/auth", strings.NewReader(val.Encode()),
   )
   if err != nil {
      return nil, err
   }
   req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
   res, err := crypto.NewTransport(hello.ClientHelloSpec).RoundTrip(req)
   if err != nil {
      return nil, err
   }
   defer res.Body.Close()
   query, err := io.ReadAll(res.Body)
   if err != nil {
      return nil, err
   }
   if val := net.ParseQuery(query); val != nil {
      return &Token{val}, nil
   }
   return nil, fmt.Errorf("parseQuery %q", query)
}

// Exchange refresh token for access token.
func (t Token) OAuth() (*OAuth, error) {
   val := make(url.Values)
   val.Set("Token", t.Get("Token"))
   val.Set("service", "oauth2:https://www.googleapis.com/auth/googleplay")
   req, err := http.NewRequest(
      "POST", origin + "/auth", strings.NewReader(val.Encode()),
   )
   if err != nil {
      return nil, err
   }
   req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
   res, err := new(http.Transport).RoundTrip(req)
   if err != nil {
      return nil, err
   }
   defer res.Body.Close()
   query, err := io.ReadAll(res.Body)
   if err != nil {
      return nil, err
   }
   if val := net.ParseQuery(query); val != nil {
      return &OAuth{val}, nil
   }
   return nil, fmt.Errorf("parseQuery %q", query)
}

type devZero struct{}

func (devZero) Read(b []byte) (int, error) {
   return len(b), nil
}

File: research/2021-11-02/gpapi/googleplay.py
from . import googleplay_pb2
import requests

BASE = "https://android.clients.google.com/"
FDFE = BASE + "fdfe/"

AUTH_URL = BASE + "auth"
CHECKIN_URL = BASE + "checkin"
CONTENT_TYPE_PROTO = "application/x-protobuf"
UPLOAD_URL = FDFE + "uploadDeviceConfig"

dev = {
   # NEED ALL THESE:
   'build.version.sdk_int': '27',
   'features': 'android.hardware.touchscreen,android.hardware.wifi',
   'gl.version': '196610',
   'hasfivewaynavigation': 'false',
   'hashardkeyboard': 'false',
   'keyboard': '1',
   'navigation': '1',
   'platforms': 'arm64-v8a,armeabi-v7a,armeabi',
   'screen.density': '420',
   'screenlayout': '2',
   'touchscreen': '3',
   'vending.version': '81031200',
}

class LoginError(Exception):

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)

class GooglePlayAPI(object):

   def __init__(self, locale="en_US", timezone="UTC", device_codename="bacon"):
        self.authSubToken = None
        self.gsfId = None
        self.dfeCookie = None

   def checkin(self, email, ac2dmToken):
      request = googleplay_pb2.AndroidCheckinRequest()
      request.version = 3
      androidCheckin = googleplay_pb2.AndroidCheckinProto()
      request.checkin.CopyFrom(androidCheckin)
      stringRequest = request.SerializeToString()
      res = requests.post(
         CHECKIN_URL,
         data=stringRequest,
         headers = {"Content-Type": CONTENT_TYPE_PROTO},
      )
      print(res.status_code, res.url)
      response = googleplay_pb2.AndroidCheckinResponse()
      response.ParseFromString(res.content)
      return response.androidId

   def uploadDeviceConfig(self, ac2dmToken):
      headers = {
         "Authorization": 'Bearer ' + ac2dmToken,
         "X-DFE-Device-Id": "{0:x}".format(self.gsfId),
         'User-Agent': 'Android-Finsky (versionCode=' + dev.get('vending.version') + ',sdk=' + dev.get('build.version.sdk_int') + ')'
      }
      deviceConfig = googleplay_pb2.DeviceConfigurationProto()
      deviceConfig.glEsVersion = int(dev['gl.version'])
      deviceConfig.hasFiveWayNavigation = (dev['hasfivewaynavigation'] == 'true')
      deviceConfig.hasHardKeyboard = (dev['hashardkeyboard'] == 'true')
      deviceConfig.keyboard = int(dev['keyboard'])
      deviceConfig.navigation = int(dev['navigation'])
      deviceConfig.screenDensity = int(dev['screen.density'])
      deviceConfig.screenLayout = int(dev['screenlayout'])
      deviceConfig.touchScreen = int(dev['touchscreen'])
      for x in dev['features'].split(","):
         deviceConfig.systemAvailableFeature.append(x)
      for x in dev['platforms'].split(","):
         deviceConfig.nativePlatform.append(x)
      upload = googleplay_pb2.UploadDeviceConfigRequest()
      upload.deviceConfiguration.CopyFrom(deviceConfig)
      stringRequest = upload.SerializeToString()
      response = requests.post(
         UPLOAD_URL,
         data=stringRequest,
         headers=headers,
      )
      print(response.status_code, response.url)
